monkey/monkey_island/setup/setup_params.py

```diff
@@ -0,0 +1,18 @@
+from dataclasses import dataclass
+
+from monkey_island.cc.server_utils.consts import (
+    DEFAULT_DATA_DIR,
+    DEFAULT_LOG_LEVEL,
+    DEFAULT_SERVER_CONFIG_PATH,
+    DEFAULT_SHOULD_SETUP_ONLY,
+    DEFAULT_START_MONGO_DB,
+)
+
+
+@dataclass
+class SetupParams:
```

Comment:
I think `SetupParams` is a bad name for this. We do setup the `data_dir` if necessary, but it's purpose is to be used at runtime as a place to store files. `log_level` isn't something we really "setup" either; it's a runtime configuration option that modifies the logger's behavior. What about `Config` or `IslandConfig`.

I think mixing `setup_only`, `server_config_path` and the other config items is a mistake.

1. Maybe `setup_only` is poorly named. Can we come up with a better name that describes what it does?
2. `log_level`, `data_dir,` and `start_mongo` are all configuration options that are stored within the server config, which is found at `server_config_path`. So this class spans at least 2 layers of the setup/config process.

Reply:
`IslandConfig` is already how we call the configuration for monkeys on the island. `Config` suggests that it's modifyable, meanwhile, `SetupParams` are not - they capture the initial inputs of the island. How about `StartupArguments`? 
1. I don't want to delve too deep into refactoring, but I'll take a look and come back to you.
2. I don't see a problem with that for now.

Reply:
I think `#2` is an indicator that this PR is really trying to solve the wrong problem. There's not a clear boundary of responsibilities between `monkey_island.py` and `main()`. This is because we're trying to configure logging early, which means we need to read the config early, which lead to a data structure that violates the single responsibility principle (i.e. `SetupParams`).

The name `SetupArguments` isn't any more clear, and you still have the problem that the class is mixing different kinds of data. It's difficult to give it a good name because it has multiple responsibilities.

I think what needs to happen is:
1. `monkey_island.py` sets up a default fail-safe logger first thing. The logger logs only to stdout.
2. `monkey_island.py` passes the `--setup-only` and `--server-config` arguments to `main.py`.
3. `main.py` parses the config file. The contents of the config file can go in a data class named something like `MonkeyIslandConfig`, since all of its members are strictly configuration options.
4. `main.py` replaces the default fail-safe logger with a fully configured logger.
5. `main.py` sets up the data directory and mongodb
6. `main.py` starts monkey island `if not --setup-only`. 

Doing it this way, `monkey_island.py` would have very minimal responsibilities. `main.py` would accept only 2 parameters, `--setup-only` and `--server-config`. Neither of these needs to be part of the config dataclass as neither of them are read from the config file.If you want those parameters 2 wrapped in their own data class, you could, but I don't think it's necessary at this point.

If you want, all of the setup logic can go in its own module called `setup` or `bootstrap` or something like that. `main.py` can pass the `MonkeyIslandConfig` dataclass object to the setup module.

Reply:
> It's difficult to give it a good name because it has multiple responsibilities.

It has a single responsibility - to establish initial island parameters. The fact that some of these parameters are file paths, bools or strings and their meaning is of no concern. We need to build a single source of truth from defaults, cmd arguments and config file contents. For now, cmd arguments and config contents don't overlap, but I don't see why `start_mongodb` can't be a cmd argument as well as a config argument in the future.

> 1. monkey_island.py sets up a default fail-safe logger first thing. The logger logs only to stdout.

Logger uses `data_dir` and `log_level`. For these we need to read config file. To read config file, we need to assess the arguments. Since we already read the config file and assessed the arguments, we can assemble Island's initial arguments from config file, cmd arguments and default parameters into a data structure. 

> 2.  monkey_island.py passes the --setup-only and --server-config arguments to main.py.
> 3. main.py parses the config file. The contents of the config file can go in a data class named something like MonkeyIslandConfig, since all of its members are strictly configuration options.

We've already parsed the config file.  I have no problem with extracting only Config options from initial arguments and putting them into `MonkeyIslandConfi` if that's used further down the road.

Reply:
Refactored regardless

Reply:
> It has a single responsibility - to establish initial island parameters.

"The single responsibility principle states that each software module should have one and only one reason to change." Determining [what qualifies as a reason to change](http://blog.cleancoder.com/uncle-bob/2014/05/08/SingleReponsibilityPrinciple.html) can be difficult, but here's what I see:

1. Add/remove/change options in the config file.
2. Add/remove/change CLI parameters.
3. Change the way configuration is stored (e.g. stored in mongodb instead of a file)


> I don't see why start_mongodb can't be a cmd argument as well as a config argument in the future.

Because doing this adds complexity and ambiguity to the interface. We need to add logic to determine what should take precedence and communicate this to the user. Since Monkey Island isn't a command-line utility, it doesn't need this kind of flexibility and the additional complexity isn't worth the cost.



>> monkey_island.py sets up a default fail-safe logger first thing. The logger logs only to stdout.

> Logger uses data_dir and log_level. For these we need to read config file.

What I'm suggesting here is we use an initial logger that doesn't depend on either and only logs to stdout, just so we can log errors in between when monkey island starts and when we can fully configure logging. This isn't necessary for this PR, though.

Reply:
Ok, understood. 
Regarding logging - let's do this after we merge data_dir changes and this PR. The more we work on these, the bigger the differences and the harder it will be to merge